File: leetcode/go/000002.add-two-numbers.go
package leetcode

// Time complexity: O(n)
// Space complexity: O(n)
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummyHead := &ListNode{}
	res := dummyHead
	cur1 := l1
	cur2 := l2
	carry := 0

	for cur1 != nil || cur2 != nil {
		val1 := 0
		if cur1 != nil {
			val1 = cur1.Val
		}
		val2 := 0
		if cur2 != nil {
			val2 = cur2.Val
		}
		sum := val1 + val2 + carry
		if sum >= 10 {
			carry = 1
			sum -= 10
		} else {
			carry = 0
		}

		res.Next = &ListNode{Val: sum}
		res = res.Next

		if cur1 != nil {
			cur1 = cur1.Next
		}
		if cur2 != nil {
			cur2 = cur2.Next
		}
	}

	if carry == 1 {
		res.Next = &ListNode{Val: 1}
	}

	return dummyHead.Next
}

File: leetcode/python/000002.add-two-numbers.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
	# Time complexity: O(m + n), m = len(l1), n - len(l2)
	# Space complexity: O(1)
	def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
		dummy_head = ListNode()
		cur = dummy_head
		carry = 0
		while l1 or l2:
			x, y = 0, 0
			if l1:
				x = l1.val
				l1 = l1.next
			if l2:
				y = l2.val
				l2 = l2.next
			sum = x + y + carry
			carry = sum // 10
			cur.next = ListNode(sum % 10)
			cur = cur.next

		if carry != 0:
			cur.next = ListNode(carry)

		return dummy_head.next
